mqtt/mqtt-broker/src/broker.rs

```diff
@@ -226,22 +283,50 @@ impl Broker {
                 "invalid protocol level received from client: {}",
                 connreq.connect().protocol_level
             );
-            let ack = proto::ConnAck {
-                session_present: false,
-                return_code: proto::ConnectReturnCode::Refused(
-                    proto::ConnectionRefusedReason::UnacceptableProtocolVersion,
-                ),
-            };
+            refuse_connection!(proto::ConnectionRefusedReason::UnacceptableProtocolVersion);
+            return Ok(());
+        }
 
-            debug!("sending connack...");
-            let event = ClientEvent::ConnAck(ack);
-            let message = Message::Client(client_id.clone(), event);
-            try_send!(connreq.handle_mut(), message);
+        // [MQTT-3.1.4-3] - The Server MAY check that the contents of the CONNECT
+        // Packet meet any further restrictions and MAY perform authentication
+        // and authorization checks. If any of these checks fail, it SHOULD send an
+        // appropriate CONNACK response with a non-zero return code as described in
+        // section 3.2 and it MUST close the Network Connection.
+        let auth_id = match self.authenticate(&connreq).await {
+            Ok(Some(auth_id)) => {
+                debug!(
+                    "client {} successfully authenticated: {}",
+                    client_id, auth_id
+                );
+                auth_id
+            }
+            Ok(None) => {
+                warn!("unable to authenticate client: {}", client_id);
+                refuse_connection!(proto::ConnectionRefusedReason::BadUserNameOrPassword);
+                return Ok(());
+            }
+            Err(e) => {
+                warn!(message = "error authenticating client: {}", error = %e);
+                refuse_connection!(proto::ConnectionRefusedReason::ServerUnavailable);
+                return Err(e);
+            }
+        };
 
-            debug!("dropping connection due to invalid protocol level");
-            let message = Message::Client(client_id, ClientEvent::DropConnection);
-            try_send!(connreq.handle_mut(), message);
-            return Ok(());
+        // Check client permissions to connect
+        match self.authorize(auth_id).await {
```

Comment:
All messages should be authorized, right? Not just the connects.?

Reply:
Correct. https://github.com/Azure/iotedge/pull/2728#discussion_r397339943

Reply:
Yes